### PR TITLE
Add settings panel

### PR DESCRIPTION
**Pull request by Uberi**
It looks like this:

screenshot from 2016-01-11 14-31-05

Settings are persistent between Firefox sessions. Can currently select mode, threshold, and clear count.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
 
 const windowUtils = require("sdk/window/utils");
 var gBrowser = windowUtils.getMostRecentBrowserWindow().getBrowser();
 var gWindow = windowUtils.getMostRecentBrowserWindow();
-var gMode = prefs.prefs.mode;
+
+// load and validate settings
+var gMode = ss.storage.mode;
+if (gMode !== "threadHangs" && gMode !== "eventLoopLags") {
+	gMode = "threadHangs";
```
Comment (mikeconley) [inline index.js]: Nit - let's stick with two space indentation. Goes for the rest of your changes as well.

File: index.js [hunk 2]
```diff
@@ -53,6 +62,33 @@ function changeState(button, aBaseSVG, aAnimateSVG = mAnimateSVG) {
   });
 }
 
+var panel = require("sdk/panel").Panel({
+  contentURL: "./panel.html",
+	contentScriptFile: "./panel.js",
```
Comment (mikeconley) [inline index.js]: Busted indentation.

File: index.js [hunk 3]
```diff
@@ -53,6 +62,33 @@ function changeState(button, aBaseSVG, aAnimateSVG = mAnimateSVG) {
...
+
+panel.on("show", function() { // this event is generated automatically by the panel upon showing
+  panel.port.emit("show", { // emit event on the panel's port so the script inside knows it's shown
+		hangThreshold: gHangThreshold,
+		mode: gMode,
+	});
+});
+panel.port.on("mode-changed", function(mode) { // this event is generated automatically by the panel upon showing
+  gMode = mode;
+	ss.storage.mode = mode;
```
Comment (mikeconley) [inline index.js]: Busted indentation here too

File: index.js [hunk 4]
```diff
@@ -53,6 +62,33 @@ function changeState(button, aBaseSVG, aAnimateSVG = mAnimateSVG) {
...
+	clearCount();
+});
+panel.port.on("hang-threshold-changed", function(hangThreshold) { // this event is generated automatically by the panel upon showing
+  gHangThreshold = hangThreshold;
+	ss.storage.hangThreshold = hangThreshold;
```
Comment (mikeconley) [inline index.js]: And here

File: index.js [hunk 5]
```diff
@@ -140,9 +171,22 @@ function numEventLoopLags() {
...
 const BADGE_COLOURS = ["red", "blue", "brown", "black"];
+let numHangsObserved = 0;
+
+function updateBadge() {
+  button.badge = (numHangs - baseNumHangs) - numHangsObserved;
+  button.badgeColor = BADGE_COLOURS[button.badge % BADGE_COLOURS.length];
+}
+
+function clearCount() {
+	baseNumHangs = numHangs;
```
Comment (mikeconley) [inline index.js]: Indentation

**Comment by mikeconley** [comment]
This looks great! Just some stylistic problems with the code, but otherwise, this looks like it'll work great.

**Comment by Uberi** [comment]
Oops, my editor was set to use tabs. I've fixed the indentation now, should I do a rebase/version bump?

**Comment by mikeconley** [comment]
Just apply the changes on top and re-push to your fork to update the pull request. I believe rebasing kills the GitHub.

**Comment by Uberi** [comment]
Just updated the package, added some docs, and a bare-bones build script. This should work as a proper extension update.

File: README.md [hunk 6]
```diff
@@ -2,11 +2,31 @@
...
+* **Minimum hang threshold**: the minimum number of milliseconds before a hang/lag is added to the count. Hangs/lags shorter than this are ignored.
+    * Note that this value is rounded up to the nearest histogram bucket. That means the actual threshold may be somewhat higher.
+* **Clear count**: reset the badge counter to 0.
+
+Development
+-----------
+
+To build the extension, use `jpm xpi`. To build an XPI for distribution, simply run `./build.sh`.
+
+When changing version numbers, make sure to update them in the following places:
```
Comment (mikeconley) [inline README.md:27]: We should probably write a build script to do this at some point.

**Comment by mikeconley** [comment]
Thanks!

**Pull request merged**